pydeseq2/_formulaic.py

```diff
@@ -0,0 +1,278 @@
+"""Helpers to interact with Formulaic Formulas, taken from pertpy
+https://github.com/scverse/pertpy/blob/main/pertpy/tools/_differential_gene_expression/_formulaic.py
+
+Some helpful definitions for working with formulaic formulas
+(e.g. `~ 0 + C(donor):treatment + np.log1p(continuous)`):
+ * A *term* refers to an expression in the formula, separated by
+   `+`, e.g. `C(donor):treatment`, or `np.log1p(continuous)`.
+ * A *variable* refers to a column of the data frame passed to formulaic, e.g. `donor`.
+ * A *factor* is the specification of how a certain variable is represented in the
+   design matrix, e.g. treatment coding with base level "A" and reduced rank.
+"""  # noqa
+
+from collections import defaultdict
+from collections.abc import Mapping
+from collections.abc import Sequence
+from dataclasses import dataclass
+from typing import Any
+
+from formulaic import FactorValues
+from formulaic import ModelSpec
+from formulaic.materializers import PandasMaterializer
+from formulaic.materializers.types import EvaluatedFactor
+from formulaic.parser.types import Factor
+from interface_meta import override
+
+
+@dataclass
+class FactorMetadata:
```

Comment:
There's also quite a bunch of test cases for the `_formulaic.py` file and the `LinearModelBase` in pertpy. Would be great if you could also port them here!